### PR TITLE
webgpu: Make staging buffer disposable and create buffer with mapping when needed

### PR DESCRIPTION
**Pull request by qjia7**
To improve the perf of uploading data, we introduced the staging buffer, which is created with mapped. To reuse the allocated staging buffers, we always keep the buffer mapped, which occupies a lot of cpu memory for a large model. That results webgpu context lost when cpu is exhausted.

In fact, we can't efficiently reuse the staging buffer since it's in an asyn method `mapAsync`. Considering this, this PR makes the staging buffer disposable. Meanwhile,  try to create the storage buffer with mapped directly if there are data needed to be uploaded. Once the uploading is done, unmap it immediately. This change resolves too many cpu memories used issue and still keeps the efficient way to upload data.

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7524)

### CODE REVIEW

**Comment by xhcao** [comment]
I am confused about the description of `disable buffer mappedAtCreation`, could you follow the WebGPU spec to optimize the description, https://www.w3.org/TR/webgpu/#:~:text=11.%20Copies-,11.1.%20Buffer%20Copies,-Buffer%20copy%20operations

**Comment by qjia7** [comment]
> I am confused about the description of `disable buffer mappedAtCreation`, could you follow the WebGPU spec to optimize the description, https://www.w3.org/TR/webgpu/#:~:text=11.%20Copies-,11.1.%20Buffer%20Copies,-Buffer%20copy%20operations

Description updated. Please take another look. Thanks.

**Review by xhcao — APPROVED** [review]
LGTM, thanks

**Review by xhcao — APPROVED** [review]
Still LGTM

File: tfjs-backend-webgpu/src/backend_webgpu.ts [hunk 1]
```diff
@@ -643,38 +643,47 @@ export class WebGPUBackend extends KernelBackend {
...
-        .resourceInfo = {size, usage: this.defaultGpuBufferUsage(), buffer};
+    let buffer;
     if (tensorData.values) {
-      const stagingBuffer = this.bufferManager.acquireUploadBuffer(
-          size, GPUBufferUsage.MAP_WRITE | GPUBufferUsage.COPY_SRC);
-      const arrayBuffer = stagingBuffer.getMappedRange();
-      if (tensorData.dtype === 'int32' || tensorData.dtype === 'bool') {
-        new Int32Array(arrayBuffer).set(tensorData.values as TypedArray);
+      buffer = this.bufferManager.acquireBuffer(
+          size, this.defaultGpuBufferUsage(), true);
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/backend_webgpu.ts:649]: It's hard to say which flags are default usages. Can we remove defaultGpuBufferUsage()?

File: tfjs-backend-webgpu/src/buffer_manager.ts [hunk 2]
```diff
@@ -26,8 +26,10 @@ export class BufferManager {
 
   constructor(private device: GPUDevice) {}
 
+  // The upload buffer is not reusable.
   acquireUploadBuffer(size: number, usage: GPUBufferUsageFlags) {
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/buffer_manager.ts:30]: We can remove this and call acquireBuffer directly.

File: tfjs-backend-webgpu/src/buffer_manager.ts [hunk 3]
```diff
@@ -84,16 +86,9 @@ export class BufferManager {
...
-      buffer: GPUBuffer, size: number, usage: GPUBufferUsageFlags) {
-    buffer.mapAsync(GPUMapMode.WRITE)
-        .then(
-            () => {
-              this.releaseBuffer(buffer, size, usage);
-            },
-            (err) => {
-                // Do nothing;
-            });
+  releaseUploadBuffer(buffer: GPUBuffer) {
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/buffer_manager.ts:89]: We can also remove this. But maybe we need a function BufferManager.dispose(buffer) to cover.

File: tfjs-backend-webgpu/src/backend_webgpu.ts (hunk 1 above)
Comment (qjia7) [reply tfjs-backend-webgpu/src/backend_webgpu.ts:649]: Removing `defaultGpuBufferUsage()` is unrelated changes for this PR. I prefer doing it in a separate PR.

File: tfjs-backend-webgpu/src/buffer_manager.ts (hunk 2 above)
Comment (qjia7) [reply tfjs-backend-webgpu/src/buffer_manager.ts:30]: `acquireBuffer`/`releaseBuffer` does many other things to manage `freeBuffers` and `usedBuffers` for reusing. They are all not needed for staging buffers. So I think the current way is more clear and efficient.

File: tfjs-backend-webgpu/src/backend_webgpu.ts (hunk 1 above)
Comment (gyagp) [reply tfjs-backend-webgpu/src/backend_webgpu.ts:649]: OK

File: tfjs-backend-webgpu/src/buffer_manager.ts (hunk 2 above)
Comment (gyagp) [reply tfjs-backend-webgpu/src/buffer_manager.ts:30]: Let's keep it as is and do changes in a later PR if needed.

**Review by gyagp — APPROVED** [review]

**Pull request merged**